libr/bin/format/elf/elf.c

```diff
@@ -2627,150 +2635,276 @@ char *Elf_(r_bin_elf_get_rpath)(ELFOBJ *bin) {
+	return get_num_relocs_dynamic (info) + get_num_relocs_sections (bin);
+}
+
+static size_t populate_relocs_record_from_dynamic(ELFOBJ *bin,
+	struct dynamic_relocation_section *info, RBinElfReloc *relocs, size_t pos) {
+	size_t size = get_size_rel_mode(info->plt_mode);
+
+	for (size_t offset = 0; offset < info->rela_size; offset += info->relaent) {
+		read_reloc (bin, relocs + pos, DT_RELA, info->addr_rela + offset - bin->baddr);
+		fix_rva_and_offset (bin, relocs + pos, info->addr_rela);
+		pos++;
 	}
-	ut8 buf[sizeof (Elf_(Rela))] = {0};
-	int res = r_buf_read_at (bin->b, offset, buf, sizeof (Elf_(Rela)));
-	if (res != sizeof (Elf_(Rela))) {
-		return -1;
+
+	for (size_t offset = 0; offset < info->rel_size; offset += info->relent) {
+		read_reloc (bin, relocs + pos, DT_REL, info->addr_rel + offset - bin->baddr);
+		fix_rva_and_offset (bin, relocs + pos, info->addr_rel);
+		pos++;
 	}
-	// TODO make a single read and work with the buffer
-	size_t i = 0;
-	if (is_rela == DT_RELA) {
-		Elf_(Rela) rela;
-#if R_BIN_ELF64
-		rela.r_offset = READ64 (buf, i);
-		rela.r_info = READ64 (buf, i);
-		rela.r_addend = READ64 (buf, i);
-#else
-		rela.r_offset = READ32 (buf, i);
-		rela.r_info = READ32 (buf, i);
-		rela.r_addend = READ32 (buf, i);
-#endif
-		r->is_rela = is_rela;
-		r->offset = rela.r_offset;
-		r->type = ELF_R_TYPE (rela.r_info);
-		r->sym = ELF_R_SYM (rela.r_info);
-		r->last = 0;
-		r->addend = rela.r_addend;
-		return sizeof (Elf_(Rela));
-	} else {
-		Elf_(Rel) rel;
-#if R_BIN_ELF64
-		rel.r_offset = READ64 (buf, i);
-		rel.r_info = READ64 (buf, i);
-#else
-		rel.r_offset = READ32 (buf, i);
-		rel.r_info = READ32 (buf, i);
-#endif
-		r->is_rela = is_rela;
-		r->offset = rel.r_offset;
-		r->type = ELF_R_TYPE (rel.r_info);
-		r->sym = ELF_R_SYM (rel.r_info);
-		r->last = 0;
-		return sizeof (Elf_(Rel));
+
+	for (size_t offset = 0; offset < info->jmprel_size; offset += size) {
+		read_reloc (bin, relocs + pos, info->plt_mode, info->addr_jmprel + offset - bin->baddr);
+		fix_rva_and_offset (bin, relocs + pos, info->addr_jmprel);
+		pos++;
 	}
-}
 
-RBinElfReloc* Elf_(r_bin_elf_get_relocs)(ELFOBJ *bin) {
-	int res, rel, rela, i, j;
-	RBinElfReloc *ret = NULL;
+	return pos;
+}
 
-	if (!bin || !bin->g_sections) {
-		return NULL;
+static size_t get_section_mode(ELFOBJ *bin, size_t pos) {
+	if (!strncmp (bin->g_sections[pos].name, ".rela.", strlen (".rela."))) {
+		return DT_RELA;
+	} else if (!strncmp (bin->g_sections[pos].name, ".rel.", strlen (".rel."))) {
+		return DT_REL;
 	}
-	size_t reloc_num = get_relocs_num (bin);
-	if (!reloc_num)	{
-		return NULL;
+
+	return 0;
+}
+
+static bool is_reloc_section(size_t rel_mode) {
+	return rel_mode == DT_REL || rel_mode == DT_RELA;
+}
+
+static bool has_valid_section_header(ELFOBJ *bin, size_t pos) {
+	return bin->g_sections[pos].info < bin->ehdr.e_shnum && bin->shdr;
+}
+
+static size_t get_next_not_analysed_offset(size_t section_offset, size_t offset,
```

Comment:
i have an hard time following this function :(

can it be simplified a bit? also, i'm not sure it's right that we check for all the `if` cases when we are really working only with one type of rel (e.g. if i'm iterating over `rela` entries, this function may still check `info->addr_jmprel - base_addr <= g_offset && g_offset < info->addr_jmprel + info->jmprel_size - base_addr`.

Reply:
The main problem is that we could have obfuscated binary like that
```
section: rel.plt
false_relocation | real rela table | jmprel table | false rel table
```
We can't trust the type of relocation in the name.
Maybe we could had an option to load or not  the relocations record from section?
It could be an start's option.